lib/Partitioner/Partitioner.cpp

```diff
@@ -569,68 +581,390 @@ void Partitioner::doPartitioning(Function *F, NodeToFunctionMap &mapping) {
+          continue;
+        }
+
+        auto *inputF = mapping[input.getNode()];
+        if (subF == inputF)
+          continue;
+
+        // Check if a DAGNode for subF's parent is created or not. If not,
+        // create one.
+        if (funcDAG.find(inputF) == funcDAG.end()) {
+          std::unique_ptr<DAGNode> subDAG = llvm::make_unique<DAGNode>();
+          subDAG->name = inputF->getName();
+          subDAG->logicalDevices = {logicalID++};
+          subDAG->backendKind = mapping.getPartitionBackendKind(inputF);
+          funcDAG[inputF] = subDAG.get();
+          nodes.push_back(std::move(subDAG));
+        }
+
+        // subF is a child of inputF, inputF is a parent of subF.
+        if (parents.find(inputF) == parents.end()) {
+          funcDAG[inputF]->children.push_back(funcDAG[subF]);
+          funcDAG[subF]->parents.push_back(funcDAG[inputF]);
+          parents.insert(inputF);
+        }
+
+        // If we've already created a placeholder for this dependence, use it.
+        auto it = placeholders.find(input);
+        if (it != placeholders.end()) {
+          N.setNthInput(inp, it->second);
+          continue;
+        }
+
+        // Create a new placeholder to represent this dependence.
+        auto *save = inputF->createSave("tmp", input);
+        auto *tmp = save->getPlaceholder();
+        placeholders[input] = tmp;
+        N.setNthInput(inp, tmp);
+      }
+    }
+  }
+
+  DAG dag;
+  dag.root = std::move(DAGRoot);
+  dag.nodes = std::move(nodes);
+  partitions_.push_back(std::move(dag));
+
+  // Update links between nodes in the cloned functions. Add placeholders (and
+  // save nodes) where a link crosses a partition boundary.
+  for (auto *subF : mapping.getPartitions()) {
+    for (auto &N : subF->getNodes()) {
+      for (int inp = 0, e = N.getNumInputs(); inp < e; inp++) {
+        auto input = N.getNthInput(inp);
+        if (isa<Storage>(input.getNode()))
+          continue;
+        // Link this node to the clone of its input.
+        auto *clone = currToNew[input.getNode()];
+        N.setNthInput(inp, NodeValue(clone, input.getResNo()));
+      }
+    }
+  }
+
+  // For all DAGNode without parents, link them to the root DAG.
+  for (auto *subF : mapping.getPartitions()) {
+    if (funcDAG[subF]->parents.size() == 0) {
+      funcDAG[subF]->parents.push_back(root);
+      root->children.push_back(funcDAG[subF]);
+    }
+  }
+}
+
+FunctionToBackendKindMapTy
+Partitioner::backendBasedPartition(Function *F,
+                                   std::vector<Backend *> backends) {
+  FunctionToBackendKindMapTy ret;
+  NodeToFunctionMap mapping;
+  std::unordered_map<Node *, BackendKind> nodeToBackendKind;
+
+  // For each node find a backend that supports it.
+  for (auto &N : F->getNodes()) {
+    for (auto &backend : backends) {
+      // Find the first backend that supports this node. The order of backends
+      // is important.
+      if (backend->shouldLower(&N) || backend->isOpSupported(N)) {
```

Comment:
Why this `shouldLower` check?

Reply:
If an op(e.g. FC)  can be lowered in the backend, it means this op is supported. If not, we will check if this op can be supported directly by this Backend.

Reply:
Seeing this in action, I don't think it makes sense actually. Backends will return true for Nodes they do not know about. E.g. if a backend does not support `SparseLengthsWeightedSum`, it will return false for `isOpSupported()`, but true for `shouldLower()` even though `SparseLengthsWeightedSum` isn't lowered to anything.

One alternative here could be that when we encounter an unsupported node, clone it, and then try to lower it. If it's not lowered then it's unsupported. Else if it is lowered then recursively do the same check on all of the nodes added during lowering. This is going to require a non-trivial refactoring though, in lowering logic/figuring out how best to clone the Node temporarily and tracking the nodes created from lowering.

Reply:
> Seeing this in action, I don't think it makes sense actually. Backends will return true for Nodes they do not know about. E.g. if a backend does not support `SparseLengthsWeightedSum`, it will return false for `isOpSupported()`, but true for `shouldLower()` even though `SparseLengthsWeightedSum` isn't lowered to anything.
> 
> One alternative here could be that when we encounter an unsupported node, clone it, and then try to lower it. If it's not lowered then it's unsupported. Else if it is lowered then recursively do the same check on all of the nodes added during lowering. This is going to require a non-trivial refactoring though, in lowering logic/figuring out how best to clone the Node temporarily and tracking the nodes created from lowering.

Thanks for pointing it out!